### PR TITLE
Use celery for downloading the blocklist URL

### PR DESCRIPTION
**Pull request by jon-betts**
For: https://github.com/hypothesis/checkmate/issues/6

This adds just enough celery config to the blocklist to run inside celery. As such it doesn't have a solution for full loading of a request context for the tasks, but we don't need one yet.

**This won't do anything without a companion commit to `h-periodic`** - As this removes the existing mechanism, we need that commit first. I'm not sure how we schedule this without getting a task pileup.

# This is very dumb

**This won't work** as written, as it downloads the file. But this will only happen on one random machine, rather than all.

### Review notes

* No tests etc yet.
* I can't see a nice way of getting this to play with `h-periodic` and `h` in dev. If the celery here is local to `checkmate` then it's not going to work right, if it has the same port, it's even worse.

### To test

* Run `CHECKMATE_BLOCKLIST_URL=<s3 url>  make services dev`
* Run a script like this:

```
from checkmate.async.tasks import sync_blocklist

sync_blocklist.delay()
```

### CODE REVIEW

**Comment by seanh** [comment]
> I'm not sure how we schedule this without getting a task pileup.

h-periodic tasks expire after a few seconds, so they don't pile up. You do have to remember to configure the `expires` setting on a per-task basis though. All the existing tasks have it. Example: https://github.com/hypothesis/h-periodic/blob/4497e6d4c64c3011e8980ff45a61e3ba3e1d8e16/h_periodic/beat.py#L11

**Comment by seanh** [comment]
> I can't see a nice way of getting this to play with h-periodic and h in dev. If the celery here is local to checkmate then it's not going to work right, if it has the same port, it's even worse.

I think we may need to enhance h-periodic to be able to talk to two different instances of Celery/Rabbit. Possibly just by having Supervisor run two different Celery Beat processes with different Celery URLs. It'd be nicer if just a single Celery Beat process could do it, if possible.

File: checkmate/async/tasks.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+from celery.utils.log import get_task_logger
+
+from checkmate.async.celery import app
+from checkmate.checker.url.blocklist import Blocklist
+
+LOG = get_task_logger(__name__)
+
+
+@app.task
```
Comment (seanh) [inline checkmate/async/tasks.py:11]:
```suggestion
@celery.task(acks_late=False)
```

As mentioned above you don't want Celery to pass the task to another / the same worker again if the worker crashes and fails to acknowledge the task. You just want it to be retried when next scheduled. So you want the task to be acknowledged early

File: conf/supervisord-dev.conf [hunk 2]
```diff
@@ -9,8 +9,8 @@ stderr_events_enabled=true
 stopsignal = KILL
 stopasgroup = true
 
-[program:refresh_blocklist]
-command=sh -c "while true; do bin/fetch-blocklist; sleep 60; done"
+[program:worker]
+command=newrelic-admin run-program celery -A checkmate.async.celery:app worker --loglevel=INFO
```
Comment (seanh) [inline conf/supervisord-dev.conf:13]: This isn't exiting cleanly anymore when you <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd> the `make dev` command in dev. It needs:

```
stopsignal = KILL
stopasgroup = true
```

at the bottom of this block. Like h has: https://github.com/hypothesis/h/blob/451f290edda91d12d2ed0df80fbc2a7794db3b73/conf/supervisord-dev.conf#L23-L24

Generally I think we want all processes (not just Celery workers) to have `stopsignal = KILL` and `stopasgroup = true`. Experimentation has showed that this means that the processes exit quickly and cleanly when <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd>'s (without printing more stuff to your terminal after your prompt has been drawn) and processes aren't left running after.

Testing it now I can reliably stop `make dev` cleanly with `stopsignal = KILL` and `stopasgroup = true` and without them it reliably fails to stop cleanly, which is the same as I found in the past with the other apps.

**Review by seanh — CHANGES_REQUESTED** [review]

File: checkmate/async/policy.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+RETRY_POLICY_QUICK = {
+    "max_retries": 2,
+    # The delay until the first retry
+    "interval_start": 0.2,
+    # How many seconds added to the interval for each retry
+    "interval_step": 0.2,
+    # Maximum number of seconds to sleep between each retry
+    "interval_max": 0.6,
+}
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply checkmate/async/policy.py:18]: It is / should be used in the Celery. This is used as the broker transport settings, which determines how hard we will try when pushing a task onto the queue.

It's not used in the sense that we don't actually push any tasks to Celery, but it seemed like a good thing to have.

File: checkmate/async/tasks.py (hunk 1 above)
Comment (jon-betts) [reply checkmate/async/tasks.py:11]: I think I prefer your suggestion above. If all of our tasks are going to be like this, we can just set the default in the celery config.

This really is `app.task`. We have made the decision to have a file called `celery` with a variable called `celery` in it in H and to also import and re-export things from the package `celery`, which makes this all quite confusing. This follows the pattern in the celery docs: https://docs.celeryproject.org/en/stable/getting-started/first-steps-with-celery.html#application

File: Makefile [hunk 4]
```diff
@@ -67,6 +70,10 @@ clean:
 	@find . -type d -name "__pycache__" -delete
 	@find . -type f -name "*.gz" -delete
 
+.PHONY: pip-compile
+pip-compile:
+	@sh requirements/compile.sh
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply Makefile:75]: :+1:

File: Makefile [hunk 5]
```diff
@@ -13,6 +13,8 @@ help:
 	@echo "make docker            Make the app's Docker image"
 	@echo "make clean             Delete development artefacts (cached files, "
 	@echo "                       dependencies, etc)"
+	@echo "make pip-compile       Compile all requirements files"
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply Makefile:16]: Yep. I was just sticking with what it was called before.

File: checkmate/async/celery.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+app = Celery("checkmate")
+app.conf.update(
+    broker_url=os.environ["CELERY_BROKER_URL"],
+    # What options should we have when sending messages to the queue?
+    broker_transport_options=RETRY_POLICY_QUICK,
+    # Tell celery where our tasks are defined
+    imports=("checkmate.async.tasks",),
+    # Acknowledge tasks after the task has executed, rather than just before
+    task_acks_late=True,
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply checkmate/async/celery.py:20]: I think if this product is going to want this the other way round for every task, then the default should be flipped. I don't think there's an inter-product default.

**Comment by jon-betts** [comment]
This is dead now

**Pull request closed without merging**